gnu_dataset/mifluz-0.26.0/man/Configuration.3

'''
''' Part of the ht://Dig package   <http://www.htdig.org/>
''' Copyright (c) 1999, 2000, 2001 The ht://Dig Group
''' For copyright details, see the file COPYING in your distribution
''' or the GNU General Public License version 2 or later
''' <http://www.gnu.org/copyleft/gpl.html>
''' 
''' 
.TH Configuration 3 local
.SH NAME
Configuration \-

reads the configuration file and manages it in memory.


.SH SYNOPSIS
.nf
.ft CW

#include <Configuration.h>

Configuration config;

ConfigDefault config_defaults = {
  { "verbose", "true" },
  { 0, 0 }
};

config.Defaults(config_defaults);

config.Read("/spare2/myconfig") ;

config.Add("sync", "false");

if(config["sync"]) ...
if(config.Value("rate") < 50) ...
if(config.Boolean("sync")) ...
.ft R
.fi

.SH DESCRIPTION

The primary purpose of the
.B Configuration
class is to parse
a configuration file and allow the application to modify the internal
data structure produced. All values are strings and are converted by the 
appropriate accessors. For instance the
.B Boolean
method will 
return numerical true (not zero) if the string either contains 
a number that is different from zero or the string
.I true
.

The
.I ConfigDefaults
type is a structure of two char pointers:
the name of the configuration attribute and it's value. The end of
the array is the first entry that contains a null pointer instead of
the attribute name. Numerical
values must be in strings. For instance:
.nf
.ft CW
ConfigDefault* config_defaults = {
  { "wordlist_compress", "true" },
  { "wordlist_page_size", "8192" },
  { 0, 0 }
};
.ft R
.fi
The additional
fields of the
.B ConfigDefault
are purely informative. 


.SH FILE FORMAT

The configuration file is a plain ASCII text file. Each line in
the file is either a comment or an attribute.
Comment lines are blank lines or lines that start with a '#'.
Attributes consist of a variable name and an associated
value:
.nf
.ft CW
<name>:<whitespace><value><newline>
.ft R
.fi

The <name> contains any alphanumeric character or
underline (_) The <value> can include any character
except newline. It also cannot start with spaces or tabs since
those are considered part of the whitespace after the colon. It
is important to keep in mind that any trailing spaces or tabs
will be included.

It is possible to split the <value> across several
lines of the configuration file by ending each line with a
backslash (\). The effect on the value is that a space is
added where the line split occurs.

A configuration file can include another file, by using the special
<name>, <tt>include</tt>. The <value> is taken as
the file name of another configuration file to be read in at
this point. If the given file name is not fully qualified, it is
taken relative to the directory in which the current configuration
file is found. Variable expansion is permitted in the file name.
Multiple include statements, and nested includes are also permitted.
.nf
.ft CW
include: common.conf
.ft R
.fi



.SH METHODS
.TP
.B   Configuration()
Constructor
.TP
.B   ~Configuration() 
Destructor
.TP
.B   void Add(const String& str)
Add configuration item
.B str
to the configuration. The value
associated with it is undefined.
.TP
.B   void Add(const String& name, const String& value)
Add configuration item
.B name
to the configuration and associate
it with
.B value
.
.TP
.B   int Remove(const String& name)
Remove the
.B name
from the configuration.
.TP
.B   void NameValueSeparators(const String& s)
Let the Configuration know how to parse name value pairs.
Each character of string
.B s
is a valid separator between
the
.I name
and the
.I value.
.TP
.B   virtual int Read(const String& filename)
Read name/value configuration pairs from the file
.B filename
.
.TP
.B   const String Find(const String& name) const
Return the value of configuration attribute
.B name
as a
.I String
.
.TP
.B   const String operator[](const String& name) const
Alias to the
.B Find
method.
.TP
.B   int Value(const String& name, int default_value = 0) const
Return the value associated with the configuration attribute
.B name
, converted to integer using the atoi(3) function.
If the attribute is not found in the configuration and 
a
.B default_value
is provided, return it.
.TP
.B   double Double(const String& name, double default_value = 0) const
Return the value associated with the configuration attribute
.B name
, converted to double using the atof(3) function.
If the attribute is not found in the configuration and 
a
.B default_value
is provided, return it.
.TP
.B   int Boolean(const String& name, int default_value = 0) const
Return 1 if the value associated to
.B name
is 
either
.B 1, yes
or
.B true
.
Return 0 if the value associated to
.B name
is 
either
.B 0, no
or
.B false
.
.TP
.B   void Defaults(const ConfigDefaults *array)
Load configuration attributes from the
.I name
and
.I value
members of the
.B array
argument.
.PP

.SH AUTHORS
Loic Dachary loic@gnu.org

The Ht://Dig group http://dev.htdig.org/


.SH SEE ALSO
htdb_dump(1), htdb_stat(1), htdb_load(1), mifluzdump(1), mifluzload(1), mifluzsearch(1), mifluzdict(1), WordContext(3), WordList(3), WordDict(3), WordListOne(3), WordKey(3), WordKeyInfo(3), WordType(3), WordDBInfo(3), WordRecordInfo(3), WordRecord(3), WordReference(3), WordCursor(3), WordCursorOne(3), WordMonitor(3), mifluz(3)



gnu_dataset/mifluz-0.26.0/mifluz/Configuration.cc
//
// Configuration.cc
//
// Configuration: This class provides an object lookup table.  Each object 
//                in the Configuration is indexed with a string.  The objects 
//                can be returned by mentioning their string index. Values may
//                include files with `/path/to/file` or other configuration
//                variables with ${variable}
//
// Part of the ht://Dig package   <http://www.htdig.org/>
// Copyright (c) 1999, 2000, 2001 The ht://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU General Public License version 2 or later 
// <http://www.gnu.org/copyleft/gpl.html>
//
// $Id: Configuration.cc,v 1.11 2001/06/29 14:14:08 loic Exp $
//

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include "Configuration.h"
#include "htString.h"
#include "ParsedString.h"

#include <stdlib.h>
#include <ctype.h>
#include <locale.h>


//*********************************************************************
// Configuration::Configuration()
//
Configuration::Configuration()
{
  separators = String("=:");
  allow_multiple = 0;
}

Configuration::Configuration(const Configuration& other) : Object()
{
  separators = other.separators;
  allow_multiple = other.allow_multiple;

  DictionaryCursor cursor;
  const char* key;
  for(other.dcGlobalVars.Start_Get(cursor); (key = other.dcGlobalVars.Get_Next(cursor));) {
    Add(key, other[key]);
  }
}

//*********************************************************************
// void Configuration::NameValueSeparators(char *s)
//
void Configuration::NameValueSeparators(const String& s)
{
    separators = s;
}


//*********************************************************************
//   Add an entry to the configuration table.
//
void Configuration::Add(const String& str_arg)
{
    const char* str = str_arg;
    String	name, value;
	
    while (str && *str)
    {
        while (isspace(*str))
            str++;
        name = 0;
        if (!isalpha(*str))
            break;
        // Some isalnum() implementations don't allow all the letters that
        // isalpha() does, e.g. accented ones.  They're not POSIX.2 compliant
        // but we won't punish them with an infinite loop...
        if (!isalnum(*str))
            break;
        while (isalnum(*str) || *str == '-' || *str == '_')
            name << *str++;

        name.lowercase();
		
        //
        // We have the name.  Let's see if we will get a value
        //
        while (isspace(*str))
            str++;
        if (!*str)
        {
            //
            // End of string.  We need to store the name as a boolean TRUE
            //
            Add(name, "true");
            return;
        }

        if (!strchr((char*)separators, *str))
        {
            //
            // We are now at a new name.  The previous one needs to be set
            // to boolean TRUE
            //
            Add(name, "true");
            continue;
        }

        //
        // We now need to deal with the value
        //
        str++;			// Skip the separator
        while (isspace(*str))
            str++;
        if (!*str)
        {
            //
            // End of string reached.  The value must be blank
            //
            Add(name, "");
            break;
        }
        value = 0;
        if (*str == '"')
        {
            //
            // Ah!  A quoted value.  This should be easy to deal with...
            // (Just kidding!)
            //
            str++;
            while (*str && *str != '"')
            {
                value << *str++;
            }
            Add(name, value);
            if (*str == '"')
                str++;
            continue;
        }
        else if (*str == '\'')
        {
            // A single quoted value.
            str++;
            while (*str && *str != '\'')
            {
                value << *str++;
            }
            Add(name, value);
            if (*str == '\'')
                str++;
            continue;
        }
        else
        {
            //
            // A non-quoted string.  This string will terminate at the
            // next blank
            //
            while (*str && !isspace(*str))
            {
                value << *str++;
            }
            Add(name, value);
            continue;
        }
    }
}


//*********************************************************************
//   Add an entry to the configuration table, without allowing variable
//   or file expansion of the value.
//
void Configuration::Add(const String& name, const String& value)
{
    String	escaped;
    const char	*s = value.get();
    while (*s)
    {
        if (strchr("$`\\", *s))
            escaped << '\\';
        escaped << *s++;
    }
    ParsedString	*ps = new ParsedString(escaped);
    dcGlobalVars.Add(name, ps);
}


//*********************************************************************
//   Add an entry to the configuration table, allowing parsing for variable
//   or file expansion of the value.
//
void Configuration::AddParsed(const String& name, const String& value)
{
    ParsedString	*ps = new ParsedString(value);
    if (strcasecmp(name, "locale") == 0)
    {
        String str(setlocale(LC_ALL, ps->get(dcGlobalVars)));
        ps->set(str);

        //
        // Set time format to standard to avoid sending If-Modified-Since
        // http headers in native format which http servers can't
        // understand
        //
        setlocale(LC_TIME, "C");
    }
    dcGlobalVars.Add(name, ps);
}


//*********************************************************************
//   Remove an entry from both the hash table and from the list of keys.
//
int Configuration::Remove(const String& name)
{
    return dcGlobalVars.Remove(name);
}


//*********************************************************************
// char *Configuration::Find(const char *name) const
//   Retrieve a variable from the configuration database.  This variable
//   will be parsed and a new String object will be returned.
//
const String Configuration::Find(const String& name) const
{
    ParsedString	*ps = (ParsedString *) dcGlobalVars[name];
    if (ps)
    {
        return ps->get(dcGlobalVars);
    }
    else
    {
#ifdef DEBUG_CONFIGURATION
        fprintf (stderr, "Could not find configuration option %s\n", (const char*)name);
#endif
        return 0;
    }
}

//*********************************************************************
Object *Configuration::Get_Object(char *name) {
return dcGlobalVars[name];
}


//*********************************************************************
//
int Configuration::Value(const String& name, int default_value) const
{
    return Find(name).as_integer(default_value);
}


//*********************************************************************
//
double Configuration::Double(const String& name, double default_value) const
{
    return Find(name).as_double(default_value);
}


//*********************************************************************
// int Configuration::Boolean(char *name, int default_value)
//
int Configuration::Boolean(const String& name, int default_value) const
{
    int		value = default_value;
    const String s = Find(name);
    if (s[0])
    {
        if (s.nocase_compare("true") == 0 ||
            s.nocase_compare("yes") == 0 ||
            s.nocase_compare("1") == 0)
            value = 1;
        else if (s.nocase_compare("false") == 0 ||
                 s.nocase_compare("no") == 0 ||
                 s.nocase_compare("0") == 0)
            value = 0;
    }

    return value;
}


//*********************************************************************
//
const String Configuration::operator[](const String& name) const
{
    return Find(name);
}


//*********************************************************************
//
int Configuration::Read(const String& filename)
{
    FILE* in = fopen((const char*)filename, "r");
 
    if(!in) {
      fprintf(stderr, "Configuration::Read: cannot open %s for reading : ", (const char*)filename);
      perror("");
      return NOTOK;
    }

#define CONFIG_BUFFER_SIZE (50*1024) 
     //
     // Make the line buffer large so that we can read long lists of start
     // URLs.
     //
     char	buffer[CONFIG_BUFFER_SIZE + 1];
     char	*current;
     String	line;
     String	name;
     char	*value;
     int        len;
     while (fgets(buffer, CONFIG_BUFFER_SIZE, in))
     {
         line << buffer;
         line.chop("\r\n");
         if (line.last() == '\\')
         {
             line.chop(1);
             continue;			// Append the next line to this one
         }
 
         current = line.get();
         if (*current == '#' || *current == '\0')
         {
             line = 0;
             continue;			// Comments and blank lines are skipped
         }
 
         name = strtok(current, ": =\t");
         value = strtok(0, "\r\n");
         if (!value)
             value = "";			// Blank value
 
         //
         // Skip any whitespace before the actual text
         //
         while (*value == ' ' || *value == '\t')
             value++;
 	len = strlen(value) - 1;
 	//
 	// Skip any whitespace after the actual text
 	//
 	while (value[len] == ' ' || value[len] == '\t')
 	  {
 	    value[len] = '\0';
 	    len--;
 	  }
 
 	if (strcasecmp((char*)name, "include") == 0)
 	{
 	    ParsedString	ps(value);
 	    String		str(ps.get(dcGlobalVars));
 	    if (str[0] != '/')		// Given file name not fully qualified
 	    {
 		str = filename;		// so strip dir. name from current one
 		len = str.lastIndexOf('/') + 1;
 		if (len > 0)
 		    str.chop(str.length() - len);
 		else
 		    str = "";		// No slash in current filename
 		str << ps.get(dcGlobalVars);
 	    }
 	    Read(str);
 	    line = 0;
 	    continue;
 	}
 
         AddParsed(name, value);
         line = 0;
     }
     fclose(in);
     return OK;
}


//*********************************************************************
// void Configuration::Defaults(ConfigDefaults *array)
//
void Configuration::Defaults(const ConfigDefaults *array)
{
    for (int i = 0; array[i].name; i++)
    {
        AddParsed(array[i].name, array[i].value);
    }
}

